java-checks/src/main/java/org/sonar/java/checks/UselessConditionCheck.java

```diff
@@ -53,18 +61,43 @@ public void scanFile(JavaFileScannerContext context) {
 
   @Override
   public void visitIfStatement(IfStatementTree tree) {
-    checkCondition(tree.condition());
-    super.visitIfStatement(tree);
+    checkIf(Arrays.asList(new ExecutionState()), tree);
+  }
+
+  private void checkIf(List<ExecutionState> states, IfStatementTree tree) {
+    ExpressionEvaluatorVisitor results = checkCondition(states, tree.condition());
+    checkNestedIf(results.trueStates, tree.thenStatement());
+    if (tree.elseStatement() != null) {
+      checkNestedIf(results.falseStates, tree.elseStatement());
+    }
+  }
+
+  private void checkNestedIf(List<ExecutionState> states, Tree tree) {
+    if (tree.is(Tree.Kind.BLOCK)) {
```

Comment:
Please do not reimplement visit of tree. 

This implementation fails to visit an if statement with an anonymous class implementation as then statement.